Sources/ArcGISToolkit/Components/FeatureFormView/FormInputs/Attachments/AttachmentCarouselItemThumbnail.swift

```diff
@@ -0,0 +1,112 @@
+// Copyright 2024 Esri
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import ArcGIS
+import SwiftUI
+
+/// The UI for a single attachment in a set of attachments within an attachment form element.
+struct AttachmentCarouselItemThumbnail: View {
+    /// The form element containing the attachment.
+    private let element: AttachmentFormElement
+    
+    /// The attachment represented by the item.
+    private let formAttachment: FormAttachment
+    
+    /// A Boolean value indicating whether the user requested the attachment's preview.
+    @State private var attachmentPreviewWasRequested = false
+    
+    /// A Boolean value indicating the user has requested that the attachment be deleted.
+    @State private var deletionWillStart: Bool = false
+    
+    /// A Boolean value indicating the user has requested that the attachment be renamed.
+    @State private var renameDialogueIsShowing = false
+    
+    /// The attachment data.
+    ///
+    /// Attachment data is provided asynchronously meaning there can be a short period where the data is
+    /// not present and we need to display alternative UI.
+    @State private var data: Data?
+    
+    /// Any error encountered while retrieving the attachment's data.
+    @State private var error: Error?
+    
+    /// The new name the user has provided for the attachment.
+    @State private var newAttachmentName = ""
+    
+    /// Creates UI to display a single form attachment.
+    /// - Parameter element: The form element containing the attachment.
+    /// - Parameter formAttachment: The attachment to be displayed.
+    init(element: AttachmentFormElement, formAttachment: FormAttachment) {
+        self.element = element
+        self.formAttachment = formAttachment
+    }
+    
+    var body: some View {
+        Group {
+            if !attachmentPreviewWasRequested {
+                Button {
+                    attachmentPreviewWasRequested = true
+                } label: {
+                    Image(systemName: "square.and.arrow.down")
+                        .font(.title3)
+                        .frame(maxWidth: .infinity, maxHeight: .infinity)
+                        .background(.thinMaterial)
+                }
+            } else if let data, let uiImage = UIImage(data: data) {
```

Comment:
`uiimage` is never cached, so is created a lot. I think it would be more efficient to cache the `UIImage` once the data is there as opposed to recreating it over and over.

Reply:
Also, this is to show the thumbnail, but you're using a full-size image. There's the `FormAttachment.createThumbnailAsync(...)` method for that.  However, that only works for image attachments. In the popups, I used the `QLThumbnailGenerator` [here](https://github.com/Esri/arcgis-maps-sdk-swift-toolkit/blob/99988c2160d2932f542d90a432718abd102b5895/Sources/ArcGISToolkit/Components/Popups/Attachments/AttachmentModel.swift#L63). It provides thumbnails for all kinds of file types, including PDFs.

You can also use the `.quickLookPreview()` to display attachments full-screen. You just need to pass the `formAttachment.filePath`, as shown [here](https://github.com/Esri/arcgis-maps-sdk-swift-toolkit/blob/99988c2160d2932f542d90a432718abd102b5895/Sources/ArcGISToolkit/Components/Popups/Attachments/AttachmentPreview.swift#L80).